openpype/hosts/maya/plugins/publish/extract_arnold_scene_source.py

```diff
@@ -0,0 +1,174 @@
+
+    label = "Extract Arnold Scene Source"
+    hosts = ["maya"]
+    families = ["ass"]
+    asciiAss = False
+
+    def process(self, instance):
+        staging_dir = self.staging_dir(instance)
+        filename = "{}.ass".format(instance.name)
+        file_path = os.path.join(staging_dir, filename)
+
+        # Mask
+        mask = arnold.AI_NODE_ALL
+
+        node_types = {
+            "options": arnold.AI_NODE_OPTIONS,
+            "camera": arnold.AI_NODE_CAMERA,
+            "light": arnold.AI_NODE_LIGHT,
+            "shape": arnold.AI_NODE_SHAPE,
+            "shader": arnold.AI_NODE_SHADER,
+            "override": arnold.AI_NODE_OVERRIDE,
+            "driver": arnold.AI_NODE_DRIVER,
+            "filter": arnold.AI_NODE_FILTER,
+            "color_manager": arnold.AI_NODE_COLOR_MANAGER,
+            "operator": arnold.AI_NODE_OPERATOR
+        }
+
+        for key in node_types.keys():
+            if instance.data.get("mask" + key.title()):
+                mask = mask ^ node_types[key]
+
+        # Motion blur
+        attribute_data = {
+            "defaultArnoldRenderOptions.motion_blur_enable": instance.data.get(
+                "motionBlur", True
+            ),
+            "defaultArnoldRenderOptions.motion_steps": instance.data.get(
+                "motionBlurKeys", 2
+            ),
+            "defaultArnoldRenderOptions.motion_frames": instance.data.get(
+                "motionBlurLength", 0.5
+            )
+        }
+
+        # Write out .ass file
+        kwargs = {
+            "filename": file_path,
+            "startFrame": instance.data.get("frameStartHandle", 1),
+            "endFrame": instance.data.get("frameEndHandle", 1),
+            "frameStep": instance.data.get("step", 1),
+            "selected": True,
+            "asciiAss": self.asciiAss,
+            "shadowLinks": True,
+            "lightLinks": True,
+            "boundingBox": True,
+            "expandProcedurals": instance.data.get("expandProcedurals", False),
+            "camera": instance.data["camera"],
+            "mask": mask
+        }
+
+        filenames = self._extract(
+            instance.data["setMembers"], attribute_data, kwargs
+        )
+
+        if "representations" not in instance.data:
+            instance.data["representations"] = []
+
+        representation = {
+            "name": "ass",
+            "ext": "ass",
+            "files": filenames if len(filenames) > 1 else filenames[0],
+            "stagingDir": staging_dir,
+            "frameStart": kwargs["startFrame"]
+        }
+
+        instance.data["representations"].append(representation)
+
+        self.log.info(
+            "Extracted instance {} to: {}".format(instance.name, staging_dir)
+        )
+
+        # Extract proxy.
+        if not instance.data.get("proxy", []):
+            return
+
+        kwargs["filename"] = file_path.replace(".ass", "_proxy.ass")
+        filenames = self._extract(
```

Comment:
Just curious: Is there any reason not to include the second publishing into the data of representations?

Reply:
I wanted to keep the original and proxy together. Think if I add them as representation, there will be conflict for destination file names since they share the same extension?

Reply:
Maybe you can try this pseudocode
`           new_rep = {
            "name": "ass",
            "ext": "ass",
            "files": filenames,
            "stagingDir": staging_dir,
            "outputName": "proxy"
        }
`

Reply:
Maybe you can try this pseduocode...
```
representation = {
            "name": "ass",
            "ext": "ass",
            "files": the_proxy_filepaths,
            "stagingDir": staging_dir,
            "outputName": "proxy"
}
```

Reply:
Or something like this
```
# .... first extraction
if "representations" not in instance.data:
    instance.data["representations"] = []

    representation = {
            "name": "ass",
            "ext": "ass",
            "files": filenames if len(filenames) > 1 else filenames[0],
            "stagingDir": staging_dir,
            "frameStart": kwargs["startFrame"]
    }

instance.data["representations"].append(representation)

# ... second extraction
proxy_representation = {
            "name": "ass",
            "ext": "ass",
            "files": the_proxy_filepaths,
            "stagingDir": staging_dir,
            "outputName": "proxy"
}
instance.data["representations"].append(proxy_representation)
```

Reply:
Nice one! Tested and it works. Will update now.